chia/_tests/core/mempool/test_mempool_manager.py

```diff
@@ -1937,3 +1939,55 @@ async def get_coin_records(coin_ids: Collection[bytes32]) -> List[CoinRecord]:
             assert result[1] != MempoolInclusionStatus.FAILED
     except ValidationError as e:
         assert e.code == expected
+
+
+@pytest.mark.anyio
+@pytest.mark.limit_consensus_modes(allowed=[ConsensusMode.HARD_FORK_2_0])
+async def test_fill_rate_block_validation(simulator_and_wallet: OldSimulatorsAndWallets) -> None:
+    """
+    This test creates a puzzle at the maximum cost limit, and generates a block
+    with it, ensuring it passes block validation.
+    """
+
+    async def send_to_mempool(full_node: FullNodeSimulator, spend_bundle: SpendBundle) -> None:
+        res = await full_node.send_transaction(wallet_protocol.SendTransaction(spend_bundle))
+        assert res is not None and ProtocolMessageTypes(res.type) == ProtocolMessageTypes.transaction_ack
+        res_parsed = wallet_protocol.TransactionAck.from_bytes(res.data)
+        assert res_parsed.status == MempoolInclusionStatus.SUCCESS.value
+
+    async def fill_mempool_with_test_sb(full_node_api: FullNodeSimulator) -> bytes32:
+        puzzle = SerializedProgram.to((1, [[ConditionOpcode.REMARK, b"1" * 12_000]]))
+        ph = puzzle.get_tree_hash()
+        for _ in range(2):
+            await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph))
+        coin_records = await full_node_api.full_node.coin_store.get_coin_records_by_puzzle_hash(False, ph)
+        # Use the farmed coin to fill the mempool with a bundle that spends it
+        coin_spend = make_spend(coin_records[0].coin, puzzle, SerializedProgram.to([]))
+        sb = SpendBundle([coin_spend], G2Element())
+        await send_to_mempool(full_node_api, sb)
+        return sb.name()
+
+    [[full_node_api], _, _] = simulator_and_wallet
+    TEST_MAX_BLOCK_CLVM_COST = uint64(144_744_040)
+    full_node_api.full_node.constants = full_node_api.full_node.constants.replace(
+        MAX_BLOCK_COST_CLVM=TEST_MAX_BLOCK_CLVM_COST
+    )
+    full_node_api.full_node.blockchain.constants = full_node_api.full_node.blockchain.constants.replace(
+        MAX_BLOCK_COST_CLVM=TEST_MAX_BLOCK_CLVM_COST
+    )
+    assert full_node_api.full_node._mempool_manager is not None
+    full_node_api.full_node._mempool_manager.mempool.mempool_info = dataclasses.replace(
+        full_node_api.full_node._mempool_manager.mempool.mempool_info,
+        max_block_clvm_cost=CLVMCost(TEST_MAX_BLOCK_CLVM_COST),
+    )
+    sb_name = await fill_mempool_with_test_sb(full_node_api)
+    mi = full_node_api.full_node.mempool_manager.get_mempool_item(sb_name)
+    assert mi is not None and mi.cost == TEST_MAX_BLOCK_CLVM_COST
+    # Farm the block to make sure we're passing block validation
+    await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(IDENTITY_PUZZLE_HASH))
```

Comment:
this doesn't seem like an "end-to-end" test. as far as I can tell, you pay all the cost to spin up a full simulation environment, but then you just interact directly with the mempool_manager. You might as well just instantiate the `MempoolManager` directly. It would be faster and you wouldn't need any hacks replacing the constants.

For example, are you sure the farmer has updated constants? What if the block that's created does in fact exceed the limit, but you wouldn't know because it's using the default constants?